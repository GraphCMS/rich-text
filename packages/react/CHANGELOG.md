# @graphcms/rich-text-react-renderer

## 0.1.1-canary.1

### Patch Changes

🐛 Bug fixes:

- Move @graphcms/rich-text-types to dependecy section to fix the error "can't resolve @graphcms/rich-text-types"

## 0.1.1-canary.0

### Patch Changes

🐛 Bug fixes:

- fix(react): not rendering list children
- fix(react): fix can't resolve @graphcms/rich-text-types
- fix(react): html and jsx tags not being rendered correctly

⚡️ Improvements:

- feat(react): accept both array and object for content
