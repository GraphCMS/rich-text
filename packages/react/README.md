# @graphcms/rich-text-react-renderer

Render Rich Text content from GraphCMS in React applications.

## ‚ö° Getting started

You can get it on npm or Yarn.

```sh
# npm
npm i @graphcms/rich-text-react-renderer

# Yarn
yarn add @graphcms/rich-text-react-renderer
```

## üî• Usage/Examples

To render the content on your application, you'll need to provide the array of elements returned from the GraphCMS API to the `RichText` component.

```tsx
import { RichText } from '@graphcms/rich-text-react-renderer';

const content = [
  {
    type: 'paragraph',
    children: [
      {
        bold: true,
        text: 'Hello World!',
      },
    ],
  },
];

const App = () => {
  return <RichText content={content} />;
};
```

The content from the example above will render:

```html
<p>
  <b>Hello world!</b>
</p>
```

### Custom elements

By default, the elements won't have any styling, despite the `IFrame`, which we designed to be responsive. But if you have, for example, a design system and wants to use your own components with styling, you can pass a `renderers` prop to the `RichText` component. Let's see an example:

```tsx
import { RichText } from '@graphcms/rich-text-react-renderer';

/**
 * const content = [ ... ]
 */

const App = () => {
  return (
    <div>
      <RichText
        children={content}
        renderers={{
          h1: ({ children }) => <h1 className="text-white">{children}</h1>,
          bold: ({ children }) => <strong>{children}</strong>,
        }}
      />
    </div>
  );
};
```

Below you can check the full list of elements you can customize, alongside the props available for each of them.

- `a`
  - `children`: ReactNode;
  - `href`: string;
  - `className`: string;
  - `rel`: string;
  - `id`: string;
  - `title`: string;
  - `openInNewTab`: boolean;
- `class`
  - `children`: ReactNode;
  - `className`: string;
- `img`
  - `src`: string;
  - `title`: string;
  - `width`: number;
  - `height`: number;
  - `mimeType`: ImageMimeTypes;
  - `altText`: string;
- `video`
  - `src`: string;
  - `title`: string;
  - `width`: number;
  - `height`: number;
- `iframe`
  - `url`: string;
  - `width`: number;
  - `height`: number;
- `h1`
  - `children`: ReactNode;
- `h2`
  - `children`: ReactNode;
- `h3`
  - `children`: ReactNode;
- `h4`
  - `children`: ReactNode;
- `h5`
  - `children`: ReactNode;
- `h6`
  - `children`: ReactNode;
- `p`
  - `children`: ReactNode;
- `ul`
  - `children`: ReactNode;
- `ol`
  - `children`: ReactNode;
- `li`
  - `children`: ReactNode;
- `table`
  - `children`: ReactNode;
- `table_head`
  - `children`: ReactNode;
- `table_body`
  - `children`: ReactNode;
- `table_row`
  - `children`: ReactNode;
- `table_cell`
  - `children`: ReactNode;
- `blockquote`
  - `children`: ReactNode;
- `bold`
  - `children`: ReactNode;
- `italic`
  - `children`: ReactNode;
- `underlined`
  - `children`: ReactNode;
- `code`
  - `children`: ReactNode;

## üìù License

Licensed under the MIT License.

---

Made with üíú by GraphCMS üëã [join our community](https://slack.graphcms.com/)!
