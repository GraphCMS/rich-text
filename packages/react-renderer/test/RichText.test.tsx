import React from 'react';
import { render } from '@testing-library/react';
import { RichText } from '@graphcms/rich-text-react-renderer';
import { RichTextContent, EmbedProps } from '@graphcms/rich-text-types';

import {
  defaultContent as content,
  imageContent,
  videoContent,
  listContent,
  iframeContent,
  inlineContent,
  emptyContent,
  embedAssetContent,
} from './content';

describe('@graphcms/rich-text-react-renderer', () => {
  it('renders content', () => {
    const { container } = render(<RichText content={content} />);

    expect(container).toMatchInlineSnapshot(`
      <div>
        <p>
          <b>
            Hello World!
          </b>
        </p>
      </div>
    `);
  });

  it('renders content correctly if received a object with children', () => {
    const contentObject: RichTextContent = {
      children: [
        {
          type: 'paragraph',
          children: [
            {
              bold: true,
              text: 'Hello World!',
            },
          ],
        },
      ],
    };

    const { container } = render(<RichText content={contentObject} />);

    expect(container).toMatchInlineSnapshot(`
      <div>
        <p>
          <b>
            Hello World!
          </b>
        </p>
      </div>
    `);
  });

  it('renders content correctly if received a object with empty children', () => {
    const { container } = render(<RichText content={emptyContent} />);

    expect(container).toMatchInlineSnapshot(`
      <div>
        <table>
          <tbody>
            <tr>
              <td>
                <p>
                  Row 1 - Col 1
                </p>
              </td>
              <td>
                <p>
                  Row 1 - Col 2
                </p>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    `);
  });

  it('renders content with custom elements', () => {
    const { container } = render(
      <RichText
        content={content}
        renderers={{
          p: ({ children }) => <p className="text-white">{children}</p>,
          bold: ({ children }) => (
            <strong className="text-black">{children}</strong>
          ),
        }}
      />
    );

    expect(container).toMatchInlineSnapshot(`
      <div>
        <p
          class="text-white"
        >
          <strong
            class="text-black"
          >
            Hello World!
          </strong>
        </p>
      </div>
    `);
  });

  it('renders inline content', () => {
    const { container } = render(<RichText content={inlineContent} />);

    expect(container).toMatchSnapshot();
  });

  it('renders inline content with custom renderers', () => {
    const { container } = render(
      <RichText
        content={inlineContent}
        renderers={{
          bold: ({ children }) => <strong>{children}</strong>,
          italic: ({ children }) => (
            <i className="italic-class" style={{ color: 'red' }}>
              {children}
            </i>
          ),
          underline: ({ children }) => <u role="button">{children} test</u>,
          code: ({ children }) => (
            <code style={{ fontStyle: 'italic' }}>{children}</code>
          ),
        }}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('renders link', () => {
    const linkContent: RichTextContent = [
      {
        type: 'link',
        id: 'test',
        rel: 'noreferrer',
        href: 'https://graphcms.com',
        title: 'GraphCMS website',
        className: 'text-white',
        openInNewTab: true,
        children: [
          {
            text: 'GraphCMS',
          },
        ],
      },
    ];

    const { container } = render(<RichText content={linkContent} />);

    expect(container).toMatchInlineSnapshot(`
      <div>
        <a
          class="text-white"
          href="https://graphcms.com"
          id="test"
          rel="noreferrer"
          target="_blank"
          title="GraphCMS website"
        >
          GraphCMS
        </a>
      </div>
    `);
  });

  it('renders iframe', () => {
    const iframeContent: RichTextContent = [
      {
        url: 'https://www.youtube.com/watch?v=Ylmd737tw5w',
        type: 'iframe',
        children: [
          {
            text: '',
          },
        ],
      },
    ];

    const { container } = render(<RichText content={iframeContent} />);

    expect(container).toMatchSnapshot();
  });

  it('renders class', () => {
    const { container } = render(<RichText content={iframeContent} />);

    expect(container).toMatchInlineSnapshot(`
      <div>
        <div
          class="test"
        >
          <p>
            wow
          </p>
        </div>
      </div>
    `);
  });

  it('renders class with custom renderer', () => {
    const { container } = render(
      <RichText
        content={iframeContent}
        renderers={{
          class: ({ children, className }) => (
            <section className={`bg-white ${className}`}>{children}</section>
          ),
        }}
      />
    );

    expect(container).toMatchInlineSnapshot(`
      <div>
        <section
          class="bg-white test"
        >
          <p>
            wow
          </p>
        </section>
      </div>
    `);
  });

  it('renders image', () => {
    const { container } = render(<RichText content={imageContent} />);

    expect(container).toMatchInlineSnapshot(`
      <div>
        <img
          alt="photo-1564631027894-5bdb17618445.jpg"
          height="1000"
          loading="lazy"
          src="https://media.graphcms.com/output=format:webp/resize=,width:667,height:1000/8xrjYm4CR721mAZ1YAoy"
          title="photo-1564631027894-5bdb17618445.jpg"
          width="667"
        />
      </div>
    `);
  });

  it('renders image with custom renderer', () => {
    const { container } = render(
      <RichText
        content={imageContent}
        renderers={{
          img: ({ src, altText }) => <img src={src} alt={altText} />,
        }}
      />
    );

    expect(container).toMatchInlineSnapshot(`
      <div>
        <img
          alt="photo-1564631027894-5bdb17618445.jpg"
          src="https://media.graphcms.com/output=format:webp/resize=,width:667,height:1000/8xrjYm4CR721mAZ1YAoy"
        />
      </div>
    `);
  });

  it('renders video', () => {
    const { container } = render(<RichText content={videoContent} />);

    expect(container).toMatchSnapshot();
  });

  it('renders lists', () => {
    const { container } = render(<RichText content={listContent} />);

    expect(container).toMatchInlineSnapshot(`
      <div>
        <ul>
          <li>
            Embroided logo
          </li>
          <li>
            Fits well
          </li>
          <li>
            Comes in black
          </li>
          <li>
            Reasonably priced
          </li>
        </ul>
      </div>
    `);
  });

  it('should render HTML and JSX tags correctly', () => {
    const content: RichTextContent = [
      { type: 'paragraph', children: [{ text: '<Test />', code: true }] },
    ];

    const { container } = render(<RichText content={content} />);

    expect(container).toHaveTextContent('<Test />');
  });

  it('should render empty text spaces', () => {
    const content: RichTextContent = [
      {
        type: 'paragraph',
        children: [
          { text: 'Sweet black ' },
          { bold: true, text: 'cap' },
          { text: ' ' },
          { text: 'with', underline: true },
          { text: ' ' },
          { text: 'embroidered', italic: true },
          { text: ' ' },
          { bold: true, text: 'GraphCMS' },
          { text: ' logo.' },
        ],
      },
    ];

    const { container } = render(<RichText content={content} />);

    expect(container).toMatchSnapshot();
  });

  it('should render embed assets', () => {
    const references = [
      {
        id: 'cknjbzowggjo90b91kjisy03a',
        url: 'https://media.graphcms.com/dsQtt0ARqO28baaXbVy9',
        mimeType: 'image/png',
      },
      {
        id: 'ckrus0f14ao760b32mz2dwvgx',
        url: 'https://media.graphcms.com/7M0lXLdCQfeIDXnT2SVS',
        mimeType: 'video/mp4',
      },
    ];

    const { container } = render(
      <RichText content={embedAssetContent} references={references} />
    );

    expect(container).toMatchSnapshot();
  });

  it(`shouldn't render embeds if id is missing in references`, () => {
    console.error = jest.fn();

    const references = [
      {
        id: '',
        url: 'https://media.graphcms.com/dsQtt0ARqO28baaXbVy9',
        mimeType: 'image/png',
      },
      {
        id: '',
        url: 'https://media.graphcms.com/7M0lXLdCQfeIDXnT2SVS',
        mimeType: 'video/mp4',
      },
    ];

    /**
     * `id` is required in `references`, but if you remove it, or if it's empty, it can't render
     */
    const { container } = render(
      <RichText content={embedAssetContent} references={references} />
    );

    expect(console.error).toHaveBeenCalledTimes(2);
    expect(container).toMatchInlineSnapshot(`<div />`);
  });

  it('should render custom embed assets', () => {
    const references = [
      {
        id: 'cknjbzowggjo90b91kjisy03a',
        url: 'https://media.graphcms.com/dsQtt0ARqO28baaXbVy9',
        mimeType: 'image/png',
      },
      {
        id: 'ckrus0f14ao760b32mz2dwvgx',
        url: 'https://media.graphcms.com/7M0lXLdCQfeIDXnT2SVS',
        mimeType: 'video/mp4',
      },
    ];

    const { container } = render(
      <RichText
        content={embedAssetContent}
        references={references}
        renderers={{
          embed: {
            Asset: () => <div>custom asset embed</div>,
          },
        }}
      />
    );

    expect(container).toMatchInlineSnapshot(`
      <div>
        <div>
          custom asset embed
        </div>
        <div>
          custom asset embed
        </div>
      </div>
    `);
  });

  it(`shouldn't render embed assets due to missing mimeType or url`, () => {
    console.error = jest.fn();

    const references = [
      {
        id: 'cknjbzowggjo90b91kjisy03a',
      },
      {
        id: 'ckrus0f14ao760b32mz2dwvgx',
      },
    ];

    const { container } = render(
      <RichText content={embedAssetContent} references={references} />
    );

    expect(console.error).toHaveBeenCalledTimes(2);
    expect(container).toMatchInlineSnapshot(`<div />`);
  });

  it('should render custom embed models', () => {
    const content: RichTextContent = [
      {
        type: 'embed',
        nodeId: 'custom_post_id',
        children: [
          {
            text: '',
          },
        ],
        nodeType: 'Post',
      },
    ];

    const references = [
      {
        id: 'custom_post_id',
        title: 'GraphCMS is awesome :rocket:',
      },
    ];

    const { container } = render(
      <RichText
        content={content}
        references={references}
        renderers={{
          embed: {
            Post: ({ title, nodeId }: EmbedProps<{ title: string }>) => {
              return (
                <div className="post">
                  <h3>{title}</h3>
                  <p>{nodeId}</p>
                </div>
              );
            },
          },
        }}
      />
    );

    expect(container).toMatchInlineSnapshot(`
      <div>
        <div
          class="post"
        >
          <h3>
            GraphCMS is awesome :rocket:
          </h3>
          <p>
            custom_post_id
          </p>
        </div>
      </div>
    `);
  });
});
