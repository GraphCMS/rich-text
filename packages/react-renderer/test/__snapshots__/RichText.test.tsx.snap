// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@graphcms/rich-text-react-renderer renders iframe 1`] = `
<div>
  <div
    style="position: relative; overflow: hidden; width: 100%; padding-top: 56.25%;"
  >
    <iframe
      allow="fullscreen"
      frameborder="0"
      loading="lazy"
      referrerpolicy="no-referrer"
      src="https://www.youtube.com/watch?v=Ylmd737tw5w"
      style="position: absolute; top: 0px; bottom: 0px; right: 0px; left: 0px; width: 100%; height: 100%;"
    />
  </div>
</div>
`;

exports[`@graphcms/rich-text-react-renderer renders inline content 1`] = `
<div>
  <p>
    <b>
      Hey, 
    </b>
    <i>
      how
    </i>
    <u>
      are
    </u>
    <code>
      you?
    </code>
  </p>
</div>
`;

exports[`@graphcms/rich-text-react-renderer renders inline content with custom renderers 1`] = `
<div>
  <p>
    <strong>
      Hey, 
    </strong>
    <i
      class="italic-class"
      style="color: red;"
    >
      how
    </i>
    <u
      role="button"
    >
      are
       test
    </u>
    <code
      style="font-style: italic;"
    >
      you?
    </code>
  </p>
</div>
`;

exports[`@graphcms/rich-text-react-renderer renders video 1`] = `
<div>
  <video
    controls=""
    height="400"
    src="https://media.graphcms.com/oWd7OYr5Q5KGRJW9ujRO"
    title="file_example_MP4_480_1_5MG.m4v"
    width="400"
  >
    <p>
      Your browser doesn't support HTML5 video. Here is a
       
      <a
        href="https://media.graphcms.com/oWd7OYr5Q5KGRJW9ujRO"
      >
        link to the video
      </a>
       instead.
    </p>
  </video>
</div>
`;

exports[`@graphcms/rich-text-react-renderer should render empty text spaces 1`] = `
<div>
  <p>
    Sweet black 
    <b>
      cap
    </b>
     
    <u>
      with
    </u>
     
    <i>
      embroidered
    </i>
     
    <b>
      GraphCMS
    </b>
     logo.
  </p>
</div>
`;

exports[`@graphcms/rich-text-react-renderer should replace all 
 in a string with <br /> elements 1`] = `
<div>
  <p>
    Hello,
    <br />
    ⁠My name is joão pedro,
    <br />
    ⁠I'm testing a bug
  </p>
</div>
`;

exports[`custom embeds and assets should render embed video, image, and audio assets 1`] = `
<div>
  <img
    loading="lazy"
    src="https://media.graphcms.com/dsQtt0ARqO28baaXbVy9"
  />
  <video
    controls=""
    height="100%"
    src="https://media.graphcms.com/7M0lXLdCQfeIDXnT2SVS"
    width="100%"
  >
    <p>
      Your browser doesn't support HTML5 video. Here is a
       
      <a
        href="https://media.graphcms.com/7M0lXLdCQfeIDXnT2SVS"
      >
        link to the video
      </a>
       instead.
    </p>
  </video>
  <audio
    controls=""
    src="https://media.graphcms.com/H9eZ7CISSBpAKxqdSwzg"
    style="display: block; max-width: 100%; height: auto;"
  >
    <p>
      Your browser doesn't support HTML5 audio. Here is a
       
      <a
        href="https://media.graphcms.com/H9eZ7CISSBpAKxqdSwzg"
      >
        link to the audio
      </a>
       instead.
    </p>
  </audio>
</div>
`;
